### PR TITLE
OCI: Fix bundle reload and activation unconditionally

### PR DESCRIPTION
**Pull request by carabasdaniel**
Using the etag in the OCI downloader to store the SHA fixes issue https://github.com/open-policy-agent/opa/issues/4637

### CODE REVIEW

File: download/oci_dowload.go [hunk 1]
```diff
@@ -100,6 +101,7 @@ func (d *OCIDownloader) ClearCache() {
 
 // TODO: this will need implementation for the OCI downloader.
```
Comment (srenatus) [inline download/oci_dowload.go:102]: [nit] Comment still accurate?

**Review by srenatus — COMMENTED** [review]
Code looks good -- just a test would be really nice. Also @ashutosh-narkar knows this code much better than I do, so I'll wait for him to have a look.

File: download/oci_dowload.go (hunk 1 above)
Comment (carabasdaniel) [reply download/oci_dowload.go]: updated

File: download/oci_dowload.go [hunk 2]
```diff
@@ -248,25 +250,38 @@ func (d *OCIDownloader) download(ctx context.Context, m metrics.Metrics) (*downl
...
-	reader := bundle.NewCustomReader(loader).WithBaseDir(d.localStorePath)
+	reader := bundle.NewCustomReader(loader).WithBaseDir(d.localStorePath).
+		WithMetrics(m).
+		WithBundleVerificationConfig(d.bvc).
+		WithBundleEtag(etag)
 	bundleInfo, err := reader.Read()
 	if err != nil {
 		return &downloaderResponse{}, fmt.Errorf("unexpected error %w", err)
 	}
+	d.SetCache(etag) // set the current etag sha to the cache
```
Comment (ashutosh-narkar) [inline download/oci_dowload.go:277]: This can be moved to the `oneShot` method once the error check is done.

File: download/oci_dowload.go [hunk 3]
```diff
@@ -248,25 +250,38 @@ func (d *OCIDownloader) download(ctx context.Context, m metrics.Metrics) (*downl
 	if tarballDescriptor.MediaType == "" {
 		return nil, fmt.Errorf("no tarball descriptor found in the layers")
 	}
-
-	bundleFilePath := filepath.Join(d.localStorePath, "blobs", "sha256", string(tarballDescriptor.Digest.Hex()))
+	etag := string(tarballDescriptor.Digest.Hex())
+	bundleFilePath := filepath.Join(d.localStorePath, "blobs", "sha256", etag)
+	// if the downloader etag sha is the same with digest of the tarball it was already loaded
+	if d.etag == etag {
```
Comment (ashutosh-narkar) [inline download/oci_dowload.go:256]: To clarify if the image has not changed we won't end up re-downloading it again, correct ?

The benefit here seems to be we won't read/activate it but what happens during download would be helpful to understand ?

Comment (carabasdaniel) [reply download/oci_dowload.go:257]: The go oras library uses the docker resolver and makes a call to retrieve the manifest of an image based on its tag. That reply contains a SHA and if that SHA is not present in the local OCI store only then it’s downloaded.

Comment (ashutosh-narkar) [reply download/oci_dowload.go:257]: Thanks for clarifying @carabasdaniel ! If we can get some tests in to verify the end-to-end flow that would be helpful.

File: test/e2e/oci/oci_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		readBuf, err = rt.GetDataWithInput("peoplefinder/POST/api/users/allowed", inputNotAllowed)
+		if err != nil {
+			t.Fatal("failed to get data from runtime")
+		}
+
+		if !strings.Contains(string(readBuf), "false") {
+			t.Fatalf("expected true but got: %s", response)
+		}
+
+		expContains := "Bundle loaded and activated successfully"
```
Comment (srenatus) [inline test/e2e/oci/oci_test.go:82]: Hmm perhaps we should also check that the second pull did *not* do this? 🤔 Maybe I'm missing something... 🔍

**Review by srenatus — COMMENTED** [review]
Thanks for adding tests! I might have just missed it -- are we testing that we _no longer_ unconditionally reload and activate?

File: test/e2e/oci/oci_test.go (hunk 4 above)
Comment (carabasdaniel) [reply test/e2e/oci/oci_test.go:87]: Rebased and added a verification that the bundle is skipped if it's the same on the pooling mechanism. I had to set the log level to debug in the test as the skip message is shown only on debug mode.

File: test/e2e/oci/oci_test.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+		if err != nil {
+			t.Fatal("failed to get data from runtime")
+		}
+		if !strings.Contains(string(readBuf), "false") {
+			t.Fatalf("expected true but got: %s", response)
+		}
+
+		time.Sleep(3 * time.Second) // wait for the downloader pooling mechanism to kick in
+		expContains := "Bundle loaded and activated successfully"
+		skipContains := "Bundle load skipped, server replied with not modified."
```
Comment (srenatus) [inline test/e2e/oci/oci_test.go:107]: hmm. not exactly on-point, but fair enough 🤔

**Review by srenatus — DISMISSED** [review]
Thank you!

**Comment by srenatus** [comment]
Could you please rebase and squash the commits? 😃

**Review by srenatus — APPROVED** [review]

**Pull request merged**